web/src/compose_validate.ts

```diff
@@ -795,3 +795,92 @@ export function convert_mentions_to_silent_in_direct_messages(
     const silent_mention_text = people.get_mention_syntax(full_name, user_id, true);
     return silent_mention_text;
 }
+
+export const SPLIT_DELIMITER = "\n\n\n";
+
+let split_messages_enabled = false;
+
+export function is_split_messages_enabled(): boolean {
+    return split_messages_enabled;
+}
+
+/**
+ * sets the state of the option to the value passed, and
+ * also in the localstorage (if supported)
+ * @param {boolean} enable Whether to enable the option
+ */
+export function set_split_messages_enabled(enable: boolean): void {
+    split_messages_enabled = enable;
+}
+
+function delimiter_index_outside_code_block(message_content: string): number {
+    let index = message_content.indexOf(SPLIT_DELIMITER);
+    while (index !== -1) {
+        if (
+            !compose_ui.position_inside_code_block(message_content, index + 1) &&
+            !compose_ui.position_inside_code_block(message_content, index + 2)
+        ) {
+            return index;
+        }
+        index = message_content.indexOf(SPLIT_DELIMITER, index + 1);
+    }
+    return -1;
+}
+
+/**
+ * Splits a message into 2 parts, if possible
+ * @param message_content the content to be split
+ * @param split_messages_enabled whether the option is enabled
+ * @returns a pair of strings: the first being the individual split message, the second being the
+ * remaining part of the message, which will be undefined when there are no further parts to send
+ */
+export function split_message(
+    message_content: string,
+    split_messages_enabled: boolean,
+): [string, string | undefined] {
+    if (!split_messages_enabled) {
+        return [message_content, undefined];
+    }
+    if (message_content.length === 0) {
+        return ["", undefined];
+    }
+    const index = delimiter_index_outside_code_block(message_content);
+    if (index === -1) {
+        return [message_content, undefined];
+    }
+    const send_content = message_content.slice(0, index);
+    const rest_content = message_content.slice(index + SPLIT_DELIMITER.length);
```

Comment:
That bug isn't caused because of this (rather it's because of an error in passing message number in show_preview), but yes, making this change helped avoid recursion to find the next non-empty message part.